### PR TITLE
[GXARCH-14] scan parent packages

### PR DESCRIPTION
**Pull request by cingaldi**
The idea of this PR is to make ownership plotter able to apply ownership information to **all** packages in a codebase

**problem**
* ownership plotter works on top of some ownership annotation. It scans methods/classes/packages for annotations and, if found, marks the class(es) as owned by a team + belongin to some context
* This mechanism breaks because of a Java language rule: packages can be annotated but annotations don't apply to subpackages. This means that code owners have to add annotation to every package in their structure 

**solution**
* introduce the concept of "annotation scanner" to create a chain of rules that look for ownership annotation associated to a given symbol (method/class)
* introduce a **ParentPackageAnnotationScanner** that looks for annotations in the parent package

So if an annotation is not found in the class/method package, the scanner goes all the way down the package hierarchy to find the first annotated package

### CODE REVIEW

File: src/main/java/com/glovoapp/ownership/ParentPackageAnnotationScanner.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import java.lang.reflect.Method;
+import java.util.Optional;
+
+public class ParentPackageAnnotationScanner<A extends Annotation> {
+
+    private Class<A> annotationClass;
+
+    public ParentPackageAnnotationScanner(Class<A> annotationClass) {
+        this.annotationClass = annotationClass;
+    }
```
Comment (Jezorko) [inline src/main/java/com/glovoapp/ownership/ParentPackageAnnotationScanner.java:14]: [nitpick] Could be replaced with [Lombok's `@RequiredArgsConstructor`](https://projectlombok.org/features/constructor) ⁽ᵃˡˢᵒ ˢᵉᵉ [ᵒᵘʳ ᵍᵘᶦᵈᵉˡᶦⁿᵉˢ](https://github.com/Glovo/backend-guidelines/blob/master/guidelines/lombok.md#requiredargsconstructor) ᶠᵒʳ ᵐᵒʳᵉ ᶦⁿᶠᵒ ᵒⁿ ʰᵒʷ ᵗʰᶦˢ ᵃⁿⁿᵒᵗᵃᵗᶦᵒⁿ ᶦˢ ᵘˢᵉᵈ ᶦⁿ ᵗʰᵉ ᵇᵃᶜᵏ⁻ᵉⁿᵈ⁾.

ᴛʜɪs ᴀᴄᴛɪᴏɴ ᴡᴀs ᴘᴀʀᴛɪᴀʟʟʏ ᴀᴜᴛᴏᴍᴀᴛᴇᴅ ʙʏ [GɪᴛHᴜʙ's ʀᴇsᴘᴏɴsᴇ ᴛᴇᴍᴘʟᴀᴛᴇs](https://github.com/settings/replies)

File: src/main/java/com/glovoapp/ownership/ParentPackageAnnotationScanner.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
+package com.glovoapp.ownership;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.AnnotatedElement;
+import java.lang.reflect.Method;
+import java.util.Optional;
+
+public class ParentPackageAnnotationScanner<A extends Annotation> {
+
+    private Class<A> annotationClass;
```
Comment (Jezorko) [inline src/main/java/com/glovoapp/ownership/ParentPackageAnnotationScanner.java:10]:
```suggestion
    private final Class<A> annotationClass;
```

File: src/main/java/com/glovoapp/ownership/ParentPackageAnnotationScanner.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
+package com.glovoapp.ownership;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.AnnotatedElement;
+import java.lang.reflect.Method;
+import java.util.Optional;
+
+public class ParentPackageAnnotationScanner<A extends Annotation> {
```
Comment (Jezorko) [inline src/main/java/com/glovoapp/ownership/ParentPackageAnnotationScanner.java:8]:
```suggestion
public final class ParentPackageAnnotationScanner<A extends Annotation> {
```

File: src/main/java/com/glovoapp/ownership/ParentPackageAnnotationScanner.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+            return Optional.of(((Class<?>) element).getPackage().getName());
+        }
+        else if (element instanceof Method) {
+            return Optional.of(((Method) element).getDeclaringClass().getPackage().getName());
+        }
+
+        return Optional.empty();
+    }
+
+    public String getSuperPackageName(String packageName) {
```
Comment (Jezorko) [inline src/main/java/com/glovoapp/ownership/ParentPackageAnnotationScanner.java:51]:
```suggestion
    private String getSuperPackageName(String packageName) {
```

File: src/test/java/com/glovoapp/ownership/ParentPackageAnnotationScannerTest.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class ParentPackageAnnotationScannerTest {
+
+    @Test
+    void getPackageName_givenClass_shouldReturnPackageName() {
+
+        var sut = new ParentPackageAnnotationScanner(ExampleOwnershipAnnotation.class);
```
Comment (Jezorko) [inline src/test/java/com/glovoapp/ownership/ParentPackageAnnotationScannerTest.java:16]: Fun fact: `sut` means `nipple` in Polish.

```suggestion
        var scanner = new ParentPackageAnnotationScanner(ExampleOwnershipAnnotation.class);
```

**Review by Jezorko — APPROVED** [review]

File: src/test/java/com/glovoapp/ownership/ParentPackageAnnotationScannerTest.java (hunk 5 above)
Comment (cingaldi) [reply src/test/java/com/glovoapp/ownership/ParentPackageAnnotationScannerTest.java]: image

**Pull request merged**